acme/acme/client.py

```diff
@@ -92,10 +95,16 @@ def register(self, new_reg=None):
         :rtype: `.RegistrationResource`
 
         """
-        new_reg = messages.NewRegistration() if new_reg is None else new_reg
-        assert isinstance(new_reg, messages.NewRegistration)
+        if hasattr(self.directory, 'new_acct'):
```

Comment:
I don't think this works in practice. I'm not familiar enough with Python & this module to point out the reason.

Against a Pebble instance I'm seeing this code break:
```
Traceback (most recent call last):
  File "./chisel2.py", line 216, in <module>
    auth_and_issue(domains)
  File "./chisel2.py", line 120, in auth_and_issue
    client = make_client(email)
  File "./chisel2.py", line 51, in make_client
    terms_of_service_agreed=True))
  File "/home/daniel/Code/certbot/acme/acme/client.py", line 104, in register
    url = self.directory.new_reg
  File "/home/daniel/Code/certbot/acme/acme/messages.py", line 199, in __getattr__
    raise AttributeError(str(error) + ': ' + name)
AttributeError: 'Directory field not found': new_reg
```

In practice adding a `pdb.set_trace()` above this line it seems like there's no `new_acct` or `new_reg` attribute on `self.directory` at this point in execution.

For reference this is the directory Pebble returned:
```
DEBUG:requests.packages.urllib3.connectionpool:http://localhost:14000 "GET /dir HTTP/1.1" 200 248
DEBUG:acme.client:Received response:
HTTP 200
Cache-Control: public, max-age=0, no-cache
Content-Type: application/json
Replay-Nonce: M2dDhbD8hqGa5sv7I1ne1w
Date: Tue, 07 Nov 2017 14:10:08 GMT
Content-Length: 248

{
   "meta": {
      "terms-of-service": "data:text/plain,Do%20what%20thou%20wilt"
   },
   "new-account": "http://localhost:14000/sign-me-up",
   "new-nonce": "http://localhost:14000/nonce-plz",
   "new-order": "http://localhost:14000/order-plz"
}
```

Reply:
Doh! Of course as soon as I left this review comment the cause became apparent. This code is checking for `new_acct`. The directory endpoint is being registered as `new_account`:
```
(Pdb) self.directory['new_account']
*** KeyError: 'Directory field not found'
(Pdb) self.directory.new_account
u'http://localhost:14000/sign-me-up'
```